source/extensions/tracers/opentelemetry/http_trace_exporter.cc

```diff
@@ -0,0 +1,80 @@
+#include "source/extensions/tracers/opentelemetry/http_trace_exporter.h"
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "source/common/common/enum_to_int.h"
+#include "source/common/common/logger.h"
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+OpenTelemetryHttpTraceExporter::OpenTelemetryHttpTraceExporter(
+    Upstream::ClusterManager& cluster_manager,
+    const envoy::config::core::v3::HttpService& http_service)
+    : cluster_manager_(cluster_manager), http_service_(http_service) {}
+
+bool OpenTelemetryHttpTraceExporter::log(const ExportTraceServiceRequest& request) {
+  std::string request_body;
+
+  const auto ok = request.SerializeToString(&request_body);
+  if (!ok) {
+    ENVOY_LOG(warn, "Error while serializing the binary proto ExportTraceServiceRequest.");
+    return false;
+  }
+
+  const auto thread_local_cluster =
+      cluster_manager_.getThreadLocalCluster(http_service_.http_uri().cluster());
+  if (thread_local_cluster == nullptr) {
+    ENVOY_LOG(error, "OTLP HTTP exporter failed: [cluster = {}] is not configured",
+              http_service_.http_uri().cluster());
+    return false;
+  }
+
+  Http::RequestMessagePtr message = Http::Utility::prepareHeaders(http_service_.http_uri());
+
+  // The request follows the OTLP HTTP specification:
+  // https://github.com/open-telemetry/opentelemetry-proto/blob/v1.0.0/docs/specification.md#otlphttp.
+  message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
+  message->headers().setReferenceContentType(Http::Headers::get().ContentTypeValues.Protobuf);
+
+  // Add all custom headers to the request.
+  for (const auto& header_value_option : http_service_.request_headers_to_add()) {
+    message->headers().setCopy(Http::LowerCaseString(header_value_option.header().key()),
+                               header_value_option.header().value());
+  }
```

Comment:
@wbpcode  Sorry, I'm not sure I get it. Are you simply explaining how it works or should I do something about it? Here we simply expect values that are provided by the user via the config and will be part of the OTLP request. Most commonly authentication header with tokens to talk to a back-end.

Reply:
Could you maybe point me to an example? That would help a lot!

Reply:
Found some examples doing things similar as:

```
request_headers_parser_(Router::HeaderParser::configure(http_service.request_headers_to_add()));
request_headers_parser_->evaluateHeaders(message->headers(), ???) 
```
I need a `stream_info` on `evaluateHeaders` but I'm not sure where/how that comes from?

Reply:
`Tracing::Driver::startSpan` has the stream info.

Reply:
> Tracing::Driver::startSpan has the stream info.

That stream info have different lifetime with exporter or pending spans.

Reply:
I think we have two options:
1. create stream info instance for the exporter or every request, just like the http health check doing. This may bring some new complexity.
  ```
  StreamInfo::StreamInfoImpl stream_info(protocol_, parent_.dispatcher_.timeSource(),
                                         local_connection_info_provider_);
  stream_info.setUpstreamInfo(std::make_shared<StreamInfo::UpstreamInfoImpl>());
  stream_info.upstreamInfo()->setUpstreamHost(host_);
  parent_.request_headers_parser_->evaluateHeaders(*request_headers, stream_info);
  ```

2. add new comment to the API to tell end users, formatter is not supported yet for current http exporter. I think this is basically acceptable for now.

Reply:
Alright, thanks for the input all! I went with option 2 for now - Modified the `opentelemetry.proto` to indicate the headers don't support any formatting like elsewhere. I used the same wording as in https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/base.proto#envoy-v3-api-msg-config-core-v3-headervalue. 

Here's a preview 
![Uploading image.png…]()

Reply:
Document is basically OK for me.

Another suggestion is that please parse these headers and values at configuration loading time to avoid copy these headers at every time you are sending span.

Reply:
I moved "parsing" the `request_headers_to_add` in the HTTP exporter constructor and kept them as a field in the exporter. Those should "live" as long the export is still alive. Then on each export, I changed from `setCopy` to `setReference` as I read in the docs that will not create copies. Since the values are stored in the HTTP exporter, all should be fine.